File: C++/relative-ranks.cpp
// Time:  O(nlogn)
// Space: O(n)

class Solution {
public:
    vector<string> findRelativeRanks(vector<int>& nums) {
        vector<int> indices(nums.size());
        iota(indices.begin(), indices.end(), 0);
        sort(indices.begin(), indices.end(),
             [&nums](int a, int b) {
                 return nums[a] > nums[b];  
             });
        
        const vector<string> ranks = {"Gold Medal", "Silver Medal", "Bronze Medal"};
        vector<string> result(nums.size());
        for (int i = 0; i < nums.size(); ++i) {
            result[indices[i]] = i > 2 ? to_string(i + 1) : ranks[i];
        }
        return result;
    }
};

File: Python/relative-ranks.py
# Time:  O(nlogn)
# Space: O(n)

class Solution(object):
    def findRelativeRanks(self, nums):
        """
        :type nums: List[int]
        :rtype: List[str]
        """
        sorted_nums = sorted(nums)[::-1]
        ranks = ["Gold Medal", "Silver Medal", "Bronze Medal"] + map(str, range(4, len(nums) + 1))
        return map(dict(zip(sorted_nums, ranks)).get, nums)

